src/ensenso.cpp

```diff
@@ -44,303 +44,502 @@ NxLibItem imageNode(NxLibItem stereo, std::optional<NxLibItem> monocular, ImageT
+
+		Result<void> set_camera_param = setNx(command.parameters()[itmCameras][0], stereo_serial_number);
+		if (!set_camera_param) return set_camera_param.error().push_description("failed to trigger");
+	}
+
+	if (monocular) {
+		Result<std::string> serial_number = monocularSerialNumber();
+		if (!serial_number) return serial_number.error().push_description("failed to trigger");
+		mono_serial_number = *serial_number;
+
+		log(fmt::format(" - {}", mono_serial_number));
+
+		Result<void> set_camera_param = setNx(command.parameters()[itmCameras][stereo ? 1 : 0], mono_serial_number);
+		if (!set_camera_param) return set_camera_param.error().push_description("failed to trigger");
+	}
+
+	Result<void> execute_trigger = executeNx(command);
+	if (!execute_trigger) return execute_trigger.error();
+
+	if (stereo) {
+		Result<bool> get_triggered = getNx<bool>(command.result()[stereo_serial_number][itmTriggered]);
+		if (!get_triggered && !*get_triggered) return get_triggered.error().push_description("failed to trigger");
 	}
 	if (monocular) {
-		std::string serial = getNx<std::string>(monocular_node.value()[itmSerialNumber]);
-		log(fmt::format(" - {}", serial));
-		setNx(command.parameters()[itmCameras][stereo ? 1 : 0], serial);
+		Result<bool> get_triggered = getNx<bool>(command.result()[mono_serial_number][itmTriggered]);
+		if (!get_triggered && !*get_triggered) return get_triggered.error().push_description("failed to trigger");
 	}
-	executeNx(command);
 
-	if (stereo && !getNx<bool>(command.result()[serialNumber()][itmTriggered])) return false;
-	if (monocular && !getNx<bool>(command.result()[monocularSerialNumber()][itmTriggered])) return false;
-	return true;
+	return estd::in_place_valid;
 }
 
-bool Ensenso::retrieve(bool trigger, unsigned int timeout, bool stereo, bool monocular) const {
+Result<void> Ensenso::retrieve(bool trigger, unsigned int timeout, bool stereo, bool monocular) const {
 	monocular = monocular && monocular_node;
 
 	// nothing to do?
-	if (!stereo && !monocular)
-		return true;
+	if (!stereo && !monocular) return estd::error("failed to retrieve: neither stereo or monucular is specified");
+
+	std::string stereo_serial_number = "";
+	std::string mono_serial_number = "";
 
 	NxLibCommand command(trigger ? cmdCapture : cmdRetrieve);
-	setNx(command.parameters()[itmTimeout], int(timeout));
-	if (stereo) setNx(command.parameters()[itmCameras][0], serialNumber());
-	if (monocular) setNx(command.parameters()[itmCameras][stereo ? 1 : 0], getNx<std::string>(monocular_node.value()[itmSerialNumber]));
-	executeNx(command);
 
-	if (stereo && !getNx<bool>(command.result()[serialNumber()][itmRetrieved])) return false;
-	if (monocular && !getNx<bool>(command.result()[monocularSerialNumber()][itmRetrieved])) return false;
-	return true;
+	Result<void> set_timeout_param = setNx(command.parameters()[itmTimeout], int(timeout));
+	if (!set_timeout_param) return set_timeout_param.error().push_description("failed to retrieve");
```

Comment:
```suggestion
	if (!set_timeout_param) return set_timeout_param.error().push_description("failed set timeout for retrieving camera data");
```

I'd try to be as specific as possible to aid in debugging both for us and the customer.